General Quarkus Platform doc describing how a platform is defined (#13153)
quarkusio/quarkus: docs/src/main/asciidoc/platform.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Platform

include::./attributes.adoc[]

The Quarkus extension ecosystem consists of the Quarkus extensions developed and maintained by the community, including the Quarkus core development team. While the Quarkus ecosystem (sometimes also referred to as the "Quarkus universe") includes all the Quarkus extensions ever developed, there is also a concept of a Quarkus platform.

== Quarkus Platform

The fundamental promise of a Quarkus platform is any combination of the Quarkus extensions the platform consists of can be used in the same application without causing any conflict for each other.

> Each organization creating their Quarkus platform may establish their own criterions for the extensions to be accepted into the platform and the means guarantee the compatibility between the accepted extensions.

Reviewer: ```suggestion
Each organization creating their Quarkus platform may establish their own criterions for the extensions to be accepted into the platform and the means to guarantee the compatibility between the accepted extensions.
```
Not entirely sure but it looks like a word is missing.

=== Quarkus Platform Artifacts

Each Quarkus platform is defined with a few artifacts.

=== Quarkus Platform BOM

Each Quarkus Platform is expected to provide a Maven BOM artifact that

> * imports a chosen version of `io.quarkus:quarkus-bom` (the platform BOM maybe flattened at the end but it has to be based on some version of `io.quarkus:quarkus-bom`)

Reviewer: ```suggestion
* imports a chosen version of `io.quarkus:quarkus-bom` (the platform BOM may be flattened at the end but it has to be based on some version of `io.quarkus:quarkus-bom`)
```

* includes all the Quarkus extension artifacts (the runtime and the deployment ones) the platform consists of

> * includes all the necessary thirdparty artifacts that align the transitive dependency versions of the platform extensions to guarantee compatiblity between them

Reviewer: ```suggestion
* includes all the necessary third-party artifacts that align the transitive dependency versions of the platform extensions to guarantee compatibility between them
```

* includes the <<platform-descriptor,platform JSON descriptor>> artifact
* possibly includes the <<platform-properties,platform configuration properties>> artifacts

Quarkus applications that want to include extensions from a Quarkus platform will be importing the Quarkus platform BOM.

[[platform-descriptor]]
=== Quarkus Platform Descriptor

Quarkus platform descriptor is a JSON artifact that provides information about the platform and its extensions to the Quarkus tools. E.g. http://code.quarkus.io and the Quarkus command line tools consult this descriptor to list, add and remove extensions to/from the project on user's request.

> This artifact is also used as a Quarkus platform identifier. When Quarkus tools need to identify the Quarkus platform(s) used in the project, they will analyze the dependency version constraints of the project (the effective list of the managed dependencies from the `dependencyManagement` section in Maven terms) looking for the platform descriptor artifact(s) among them. Given that the platform descriptors are included into the Quarkus platform BOMs, every Quarkus application will inherit the platform descriptor artifact from the imported platform BOM(s) as a dependency version constraint (managed depenency in Maven terms).

Reviewer: ```suggestion
This artifact is also used as a Quarkus platform identifier. When Quarkus tools need to identify the Quarkus platform(s) used in the project, they will analyze the dependency version constraints of the project (the effective list of the managed dependencies from the `dependencyManagement` section in Maven terms) looking for the platform descriptor artifact(s) among them. Given that the platform descriptors are included into the Quarkus platform BOMs, every Quarkus application will inherit the platform descriptor artifact from the imported platform BOM(s) as a dependency version constraint (managed dependency in Maven terms).
```

> To be able to easily identify Quarkus platform descriptors among the project's dependency constraints, the platform descriptor Maven artifact coordinates should follow the folloing naming convention:

Reviewer: ```suggestion
To be able to easily identify Quarkus platform descriptors among the project's dependency constraints, the platform descriptor Maven artifact coordinates should follow the following naming convention:
```

* the `groupId` of the descriptor artifact should match the `groupId` of the corresponding Quarkus Platform BOM;
* the `artifactId` of the descriptor artifact should be the `artifactId` of the corresponding Quarkus Platform BOM with the `-quarkus-platform-descriptor` suffix;

> * the `classifier` of the descriptor artifact should match the `version` of the corresponding Quarkus Platform BOM;

Reviewer: Ah ah I added a question here but you have a nice answer below :).

* the `type` of the descriptor artifact should be `json`;
* the `version` of the descriptor artifact should match the `version` of the  corresponding Quarkus Platform BOM.

> As a string it will look as `<platform-bom-groupId>:<platform-bom-artifactId>-quarkus-platform-descriptor:<platform-version>:json:<platform-version>`

Reviewer: ```suggestion
As a string it will look like `<platform-bom-groupId>:<platform-bom-artifactId>-quarkus-platform-descriptor:<platform-version>:json:<platform-version>`
```

  <configuration>
    <bomGroupId>${quarkus.platform.group-id}</bomGroupId> <3>
    <bomArtifactId>${quarkus.platform.artifact-id}</bomArtifactId> <4>
    <bomVersion>${quarkus.platform.version}</bomVersion> <5>
    <overridesFile>${overridesfile}</overridesFile> <6>
    <resolveDependencyManagement>true</resolveDependencyManagement> <7>
  </configuration>
</plugin>
----

<1> the version of the `quarkus-platform-descriptor-json-plugin`
<2> `generate-extensions-json` is the goal generating the platform descriptor
<3> the `groupId` of the platform BOM
<4> the `artifactId` of the platform BOM
<5> the `version` of the platform BOM
<6> this parameter is optional, it allows to override some metadata from the Quarkus extension descriptors found in every runtime extension artifact from which the platform descriptor is generated
<7> this parameter is also optional and defaults to false. It has to be set to true in case the platform BOM *is not generated* and *is not flattened*. Which for example is the case for `io.quarkus:quarkus-bom`.

[[platform-properties]]
=== Quarkus Platform Properties

A Quarkus platform may provide its own default values for some of the configuration options.

> Quarkus is using https://github.com/eclipse/microprofile-config[Eclipse MicroProfile Config^] for wiring application configuration. A Quarkus platform may be used as another source of configuration in the hierarchy of the configuration sources dominated by the application's `application.properties`.

Reviewer: ```suggestion
Quarkus is using https://github.com/eclipse/microprofile-config[Eclipse MicroProfile Config] for wiring application configuration. A Quarkus platform may be used as another source of configuration in the hierarchy of the configuration sources dominated by the application's `application.properties`.
```

To provide platform-specific defaults, the platform needs to include a dependency version constraint in its BOM for a properties artifact whose coordinates follow the following naming convention:

* the `groupId` of the properties artifact should match the `groupId` of the corresponding Quarkus Platform BOM;
* the `artifactId` of the properties artifact should be the `artifactId` of the corresponding Quarkus Platform BOM with the `-quarkus-platform-properties` suffix;
* the `classifier` of the descriptor artifact should be left empty/null;
* the `type` of the descriptor artifact should be `properties`;
* the `version` of the descriptor artifact should match the `version` of the  corresponding Quarkus Platform BOM.

The properties artifact itself is expected to be a traditional `properties` file that will be loaded into an instance of `java.util.Properties` class.

> IMPORTANT: At this point, platform properties are only allowed to provide the default values to a restricted set of properties. Property names in the platform properties file must be prefixed with the `platform.` suffix.

Reviewer: ```suggestion
IMPORTANT: At this point, platform properties are only allowed to provide the default values of a restricted set of properties. Property names in the platform properties file must be prefixed with the `platform.` suffix.
```
Reads better IMHO.

Author: Perhaps "for" will be even better?

Author: I think "provide ... for" sounds better. And probably it should be "a restricted set of configuration options".

Author: I re-wrote it as
-IMPORTANT: At this point, platform properties are only allowed to provide the default values to a restricted set of properties. Property names in the platform properties file must be prefixed with the `platform.` suffix.
+IMPORTANT: At this point, platform properties are only allowed to provide the default values for a restricted set of configuration options. The property names in the platform properties file must be prefixed with the `platform.` suffix.

Thanks for the review!

Extension developers that want to make their configuration options platform-specific should set their default values to properties that start with the `platform.` suffix. Here is an example:

[source,java]
----
package io.quarkus.deployment.pkg;

@ConfigRoot(phase = ConfigPhase.BUILD_TIME)
public class NativeConfig {

    /**
     * The docker image to use to do the image build
     */
    @ConfigItem(defaultValue = "${platform.quarkus.native.builder-image}")
    public String builderImage;
}
----

In this case the default value for `quarkus.native.builder-image` will be provided by the platform. The user will still be able to set the desired value for `quarkus.native.builder-image` in its `application.properties`, of course. But in case it's not customized by the user, the default value will be coming from the platform properties.

> A platform properties file for the example above would contain (the actual value is provided as an example)

Reviewer: ```suggestion
A platform properties file for the example above would contain (the actual value is provided as an example):
```

[source,text]
----

> platform.quarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-native-image:20.2.0-java11

Reviewer 2: Looking at [`platform-properties.properties`](https://github.com/quarkusio/quarkus/blob/master/devtools/platform-properties/src/main/resources/platform-properties.properties) provided by Quarkus I get the impression that it is also possible to have the values refer to a Maven property by the same (or a different) name. If that is how it works, should then this also be described in here?

Author: Do you mean `platform.quarkus.xxx` properties? If so, I would not recommend users to set `platform.quarkus.xxx` properties. It will end up being confusing. `quarkus.xxx` properties are the proper way to customize the config.

Perhaps, there is a use-case I didn't think of. In which case, please, clarify why you would prefer customizing the default config not using the `quarkus.xxx` properties.

`platform.quarkus.xxx` and `quarkus.xxx` properties are on different levels in the ConfigSource hierarchy.
E.g. if you add a `quarkus.xxx` property to the `application.properties`, it can still be overridden on the command line with `-Dquarkus.xxx` for example. But `-Dplatform.quarkus.xxx` won't override it.
However, if instead of `quarkus.xxx` property you add `platform.quarkus.xxx` to the `application.properties` then you would be able to override it on the command line with both `-Dquarkus.xxx` and `-Dplatform.quarkus.xxx`.
Perhaps we should look into restricting the `platform.quarkus.xxx` properties further. I would like users to set them directly. This was not intended to be supported.

Reviewer 2: In our project we intend to build a Quarkus platform directly building upon what you are describing here, so I am trying to better understand how this works by looking at the implementation.

In the file I refer to you have `platform.quarkus.native.builder-image=${platform.quarkus.native.builder-image}` which confuses me a bit, as it looks like a recursive definition. The only other place I found this property being referenced was in [build-parent/pom.xml](https://github.com/quarkusio/quarkus/blob/master/build-parent/pom.xml#L128):
```xml
<platform.quarkus.native.builder-image>quay.io/quarkus/ubi-quarkus-native-image:20.2.0-java11</platform.quarkus.native.builder-image>
```

I should probably look a biut further into the code, but in this documentation I couldn't find this mechanism described.

Author: It's very simple. This file https://github.com/quarkusio/quarkus/blob/master/devtools/platform-properties/src/main/resources/platform-properties.properties is located under `src/main/resources` which is processed (filtered) by the Maven resources plugin and copied to the `target/classes` directory with the `${xxx}` replaced with the actual value set in the `build-parent/pom.xml`.
It's just instead of configuring the defaults right in the platform-properties.properties file itself, I decided to provide the value in the pom.xml. There is no recursive definition. I could have called the POM property differently, but it looked simpler and less ambiguous to me to use the same name.

Reviewer 2: I see. Thanks for the explanation! That makes perfect sense and seems like a good way to define the platform.

Now I suppose Quarkus will over time gain some more platform properties. For our platform we would like to set many more of the Quarkus properties in the platform. I think for that purpose we will try to define a config source which sits between the defaults provided by the config items and the application's `application.properties`. I assume this should at least work for runtime properties. Once I have worked out which these (one example would be `quarkus.log.console.format`) I will try to file corresponding tickets.

Author: Yes, please, do let us know what issues you are facing. Thanks!

Reviewer 2: One thing I would like to check up on is whether the all-config documentation generation still works correctly after this change. I am thinking that the following could possibly cause problems. Or does the `extensions.json` get copied into `bom-descriptor-json` as well?

https://github.com/quarkusio/quarkus/blob/a0374b7517c859bb349a06d4c1264f185d345e7d/docs/pom.xml#L96

Author: Thanks for the PR fixing the path. The way the current docs are generated, the value for the config options configured as platform properties instead of containing actual values will appear as `${platform.quarkus.xxx}`.

----

> There is also a Maven plugin goal that validates the platform properties content and its artifact coordinates and also checks whether the platform properties artifact is present in the platform's BOM. Here is a sample plugin configuration

Reviewer: ```suggestion
There is also a Maven plugin goal that validates the platform properties content and its artifact coordinates and also checks whether the platform properties artifact is present in the platform's BOM. Here is a sample plugin configuration:
```
